core/vdbe/execute.rs

```diff
@@ -4443,79 +4443,136 @@ pub fn op_idx_delete(
+            let CursorType::BTreeIndex(index_meta) = cursor_type else {
+                panic!("IdxInsert: not a BTreeIndex cursor");
             };
-            // To make this reentrant in case of `moved_before` = false, we need to check if the previous cursor.insert started
-            // a write/balancing operation. If it did, it means we already moved to the place we wanted.
-            let moved_before = if cursor.is_write_in_progress() {
-                true
-            } else if index_meta.unique {
-                // check for uniqueness violation
-                match cursor.key_exists_in_index(record)? {
-                    CursorResult::Ok(true) => {
-                        if flags.has(IdxInsertFlags::NO_OP_DUPLICATE) {
-                            break 'block;
-                        }
-                        return Err(LimboError::Constraint(
-                            "UNIQUE constraint failed: duplicate key".into(),
-                        ));
-                    }
-                    CursorResult::IO => return Ok(InsnFunctionStepResult::IO),
-                    CursorResult::Ok(false) => {}
+            if !index_meta.unique {
+                state.op_idx_insert_state = OpIdxInsertState::Insert {
+                    moved_before: false,
                 };
-                // uniqueness check already moved us to the correct place in the index.
-                // the uniqueness check uses SeekOp::GE, which means a non-matching entry
-                // will now be positioned at the insertion point where there currently is
-                // a) nothing, or
-                // b) the first entry greater than the key we are inserting.
-                // In both cases, we can insert the new entry without moving again.
-                //
-                // This is re-entrant, because once we call cursor.insert() with moved_before=true,
-                // we will immediately set BTreeCursor::state to CursorState::Write(WriteInfo::new()),
-                // in BTreeCursor::insert_into_page; thus, if this function is called again,
-                // moved_before will again be true due to cursor.is_write_in_progress() returning true.
-                true
+                return Ok(InsnFunctionStepResult::Step);
+            }
+            {
+                let mut cursor = state.get_cursor(cursor_id);
+                let cursor = cursor.as_btree_mut();
+
+                return_if_io!(cursor.seek(
+                    SeekKey::IndexKey(record_to_insert),
+                    SeekOp::GE { eq_only: true }
+                ));
+            }
+            state.op_idx_insert_state = OpIdxInsertState::UniqueConstraintCheck;
+            Ok(InsnFunctionStepResult::Step)
+        }
+        OpIdxInsertState::UniqueConstraintCheck => {
+            let ignore_conflict = 'i: {
+                let mut cursor = state.get_cursor(cursor_id);
```

Comment:
what is this `'i: {` magic? how does it even break a non loop block? It just exits it?

Reply:
yes! you can break out of it and assign a value